flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/proxy/KinesisProxyV2.java

```diff
@@ -28,16 +51,210 @@
+						fanOutStreamInfo = describeStreamConsumer(stream, streamArn, consumerName);
+					} else if (AwsV2Util.isRecoverableException(ex)) {
+						long backoffMillis = fullJitterBackoff(
+							fanOutRecordPublisherConfiguration.getRegisterStreamBaseBackoffMillis(), fanOutRecordPublisherConfiguration.getRegisterStreamMaxBackoffMillis(), fanOutRecordPublisherConfiguration.getRegisterStreamExpConstant(), retryCount++);
+						LOG.warn("Got recoverable AmazonServiceException when trying to register " + stream + ". Backing off for "
+							+ backoffMillis + " millis (" + ex.getClass().getName() + ": " + ex.getMessage() + ")");
+						Thread.sleep(backoffMillis);
+					} else {
+						throw ex;
+					}
+				}
+			}
+
+			if (fanOutStreamInfo == null) {
+				throw new RuntimeException("Retries exceeded for registerStream operation - all " + fanOutRecordPublisherConfiguration.getRegisterStreamMaxRetries() +
+					" retry attempts failed.");
+			}
+			result.add(fanOutStreamInfo);
+		}
+		return result;
+	}
+
+	public FanOutStreamInfo describeStreamConsumer(String stream, String streamArn, String consumerName) throws InterruptedException, ExecutionException  {
+		DescribeStreamConsumerRequest describeStreamConsumerRequest = DescribeStreamConsumerRequest
+			.builder()
+			.streamARN(streamArn)
+			.consumerName(consumerName)
+			.build();
+		FanOutStreamInfo fanOutStreamInfo = null;
+		int retryCount = 0;
+		while (retryCount <= fanOutRecordPublisherConfiguration.getDescribeStreamConsumerMaxRetries() && fanOutStreamInfo == null) {
+			try {
+				DescribeStreamConsumerResponse describeStreamConsumerResponse = kinesisAsyncClient.describeStreamConsumer(describeStreamConsumerRequest).get();
+				fanOutStreamInfo = new FanOutStreamInfo(stream, streamArn, consumerName, describeStreamConsumerResponse.consumerDescription().consumerARN());
+			} catch (ExecutionException ex) {
+				if (AwsV2Util.isRecoverableException(ex)) {
+					long backoffMillis = fullJitterBackoff(
+						fanOutRecordPublisherConfiguration.getDescribeStreamConsumerBaseBackoffMillis(), fanOutRecordPublisherConfiguration.getDescribeStreamConsumerMaxBackoffMillis(), fanOutRecordPublisherConfiguration.getDescribeStreamConsumerExpConstant(), retryCount++);
+					LOG.warn("Got recoverable AmazonServiceException when trying to describe stream consumer " + stream + ". Backing off for "
+						+ backoffMillis + " millis (" + ex.getClass().getName() + ": " + ex.getMessage() + ")");
+					Thread.sleep(backoffMillis);
+				} else {
+					throw ex;
+				}
+			}
+		}
+
+		if (fanOutStreamInfo == null) {
+			throw new RuntimeException("Retries exceeded for registerStream operation - all " + fanOutRecordPublisherConfiguration.getRegisterStreamMaxRetries() +
+				" retry attempts failed.");
+		}
+		return fanOutStreamInfo;
 	}
 
+	/**
+	 * {@inheritDoc}
+	 */
+	@Override
+	public void deregisterStreamConsumer(List<FanOutStreamInfo> fanOutStreamInfos) throws InterruptedException, ExecutionException {
```

Comment:
@xiaolong-sn again please make this singular (single deregistration at a time)

Also this needs to open with describe and stagger too as per the FLIP flow chart to deal with competing requests

Reply:
About `deregisterStreamConsumer` operation, why can’t we simply deregister first, and if the exception it throws is `ResourceNotFoundException` , we then simply end the operation ( for it indicates that the consumer is already deregistered by others).

That way, we do not have to call the `describeStreamConsumer` operation at all.

Reply:
I changed the behavior and interface.